Modified the README.md to include a second example

asparks1805 opened the pull request:
Modified the README.md to include a second example where the account SID and
auth token are embedded in the code rather than stored as environmental
variables. This would allow a developer to compile a self-contained executable
that could be run by end users without the need to create or understand
environment variables. Also modified both examples to define sending and
receiving phone number bindings to make it easier for the developer to
read/configure.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/markandrus/twilio-haskell/40)

<!-- Reviewable:end -->

asparks1805 committed:
Modified the README.md to include a second example

Modified the README.md to include a second example where the account SID and
auth token are embedded in the code rather than stored as environmental
variables. This would allow a developer to compile a self-contained executable
that could be run by end users without the need to create or understand
environment variables. Also modified both examples to define sending and
receiving phone number bindings to make it easier for the developer to
read/configure.

markandrus commented on README.md:
```diff
@@ -38,11 +38,53 @@ main = runTwilio' (getEnv "ACCOUNT_SID")
-  let body = PostMessage "+14158059869" "+14158059869" "Oh, hai"
+  let receivingPhone = "+14158059869"
+  let sendingPhone = "+14158059869"
+  let body = PostMessage receivingPhone sendingPhone "Oh, hai"
```
What do you think about using the names `to` and `from` instead of `receivingPhone` and `sendingPhone`? This would better match the REST API.

asparks1805 replied:
## Yep, works for me. I'll make the change and update the pull request tomorrow morning. 

Aaron

markandrus commented on README.md:
```diff
@@ -38,11 +38,53 @@ main = runTwilio' (getEnv "ACCOUNT_SID")
+import Control.Monad.IO.Class (liftIO)
+import Twilio
+import Twilio.Messages
+import qualified Data.Text as T
```
Since only `Text` is being used, what do you think about going with a simpler import, e.g.

``` hs
import Data.Text (Text)
```

Do you mind alphabetizing the imports as well?

markandrus commented on README.md:
```diff
@@ -38,11 +38,53 @@ main = runTwilio' (getEnv "ACCOUNT_SID")
+                Just accountSID ->
+                    Just (accountSID, authToken)
+                Nothing -> Nothing)
+        Nothing -> Nothing
```
Here you are pattern matching on the success/failure of the parses to determine the next step. Instead of explicit pattern matching, we can rely on the behavior of Maybe's Functor and Applicative instances to write this more succinctly. For example,

``` hs
parseCredentials :: Text -> Text -> Maybe Credentials
parseCredentials accountSID authToken =
  (,) <$> parseAccountSID accountSID <*> parseAuthToken authToken
```

I wonder if such a function should be include the library proper.

asparks1805 replied:
Yes, agreed. We saw that you actually have this function built but haven't exported it. 

## 

Aaron

markandrus replied:
Ah, you're right! And it was written with a less-palatable combination of `uncurry` and `liftM2`... the `<$>`/`<*>` combination feels nicer, IMO.

markandrus commented on README.md:
```diff
@@ -38,11 +38,53 @@ main = runTwilio' (getEnv "ACCOUNT_SID")
+                let body = PostMessage receivingPhone sendingPhone "Oh, hai"
+                message <- post body
+                liftIO $ print message
+        Nothing -> print "Something bad happened, you have poorly formed credentials."
```
Do you mind using a 2-space indent?

asparks1805 committed:
Cleared the review notes for commit 2afd62a

Cleared the review notes for commit 2afd62a (modified README for a second
example). Modified the following:
* Used to/from on phone bindings to match with Twilio's RESTful API
* Alphabetized the imports
* Used simpler Data.Text (Text) rather than qualified
* Switched to 2 space indent from 4

StevenXL commented:
@markandrus Hi Mark. It seems like this PR is over a year old. Would like to know if you want to update it and try to merge it into master, or if it is a good candidate for closing. 

(P.S., haven't actually looked at it, just the date it was opened).

markandrus commented:
Sorry, I kind of dropped the ball on @asparks1805 PR. Some notes:

* @asparks1805 identified a useful function, `parseCredentials`, that we hadn't exported. It'd probably be useful to export this.
* I don't think we should land the "using SID and secret embedded in the code" example, since this is kind of an anti-pattern (someone could extract the credentials from the binary). Twilio's other examples prefer environment variables here.
* Otherwise, I do think binding the variable names in the example is useful (e.g., `toPhone`, etc.)—I get the order confused whenever I re-read it.

I'm also reminded of [this suggestion](https://github.com/markandrus/twilio-haskell/issues/58) to improve some of our examples.

I don't think we should merge this as-is. Maybe we can run with some of the bullet points above, though?
